Issue parsing Scala 3 and macros

**Describe the bug**

I'm not 100% sure what's going on as I didn't dive too deep into this, but wanted to capture what @keynmol mentioned on Gitter here.

Either there is an issue with Scalameta parsing this or we aren't correctly identifying that the source we are in is Scala 3. However, I do see that in general it recognizes that the build target is Scala 3, so I don't think that's the issue. Basically when you open a specific file with a macro in it you get a giant error and in the logs you see something along the lines of:

```
SEVERE: Internal error: /Users/ckipp/Documents/scala-workspace/expecty/src/main/scala-3/com/eed3si9n/expecty/RecorderMacro.scala:32: error: unclosed character literal
    '{
    ^
java.util.concurrent.CompletionException: /Users/ckipp/Documents/scala-workspace/expecty/src/main/scala-3/com/eed3si9n/expecty/RecorderMacro.scala:32: error: unclosed character literal
    '{
    ^
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
        at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:704)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
        at scala.meta.internal.metals.CancelTokens$.$anonfun$future$1(CancelTokens.scala:40)
        at scala.meta.internal.metals.CancelTokens$.$anonfun$future$1$adapted(CancelTokens.scala:38)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: /Users/ckipp/Documents/scala-workspace/expecty/src/main/scala-3/com/eed3si9n/expecty/RecorderMacro.scala:32: error: unclosed character literal
    '{
    ^
        at scala.meta.internal.tokenizers.Reporter.syntaxError(Reporter.scala:23)
        at scala.meta.internal.tokenizers.Reporter.syntaxError$(Reporter.scala:23)
        at scala.meta.internal.tokenizers.Reporter$$anon$1.syntaxError(Reporter.scala:33)
        at scala.meta.internal.tokenizers.Reporter.syntaxError(Reporter.scala:25)
        at scala.meta.internal.tokenizers.Reporter.syntaxError$(Reporter.scala:25)
        at scala.meta.internal.tokenizers.Reporter$$anon$1.syntaxError(Reporter.scala:33)
        at scala.meta.internal.tokenizers.LegacyScanner.fetchSingleQuote$1(LegacyScanner.scala:469)
        at scala.meta.internal.tokenizers.LegacyScanner.fetchToken(LegacyScanner.scala:474)
        at scala.meta.internal.tokenizers.LegacyScanner.nextToken(LegacyScanner.scala:251)
        at scala.meta.internal.mtags.ScalaToplevelMtags.acceptBalancedDelimeters(ScalaToplevelMtags.scala:274)
        at scala.meta.internal.mtags.ScalaToplevelMtags.loop(ScalaToplevelMtags.scala:172)
        at scala.meta.internal.mtags.ScalaToplevelMtags.indexRoot(ScalaToplevelMtags.scala:50)
        at scala.meta.internal.mtags.MtagsIndexer.index(MtagsIndexer.scala:19)
        at scala.meta.internal.mtags.MtagsIndexer.index$(MtagsIndexer.scala:18)
        at scala.meta.internal.mtags.ScalaToplevelMtags.index(ScalaToplevelMtags.scala:40)
        at scala.meta.internal.mtags.Mtags$.allToplevels(Mtags.scala:96)
        at scala.meta.internal.metals.DefinitionProvider.fromMtags(DefinitionProvider.scala:212)
        at scala.meta.internal.metals.DefinitionProvider.$anonfun$positionOccurrence$3(DefinitionProvider.scala:174)
        at scala.Option.orElse(Option.scala:447)
        at scala.meta.internal.metals.DefinitionProvider.$anonfun$positionOccurrence$1(DefinitionProvider.scala:174)
        at scala.Option.flatMap(Option.scala:271)
        at scala.meta.internal.metals.DefinitionProvider.positionOccurrence(DefinitionProvider.scala:169)
        at scala.meta.internal.metals.DocumentHighlightProvider.$anonfun$documentHighlight$1(DocumentHighlightProvider.scala:33)
        at scala.collection.immutable.List.map(List.scala:293)
        at scala.meta.internal.metals.DocumentHighlightProvider.documentHighlight(DocumentHighlightProvider.scala:29)
        at scala.meta.internal.metals.MetalsLanguageServer.$anonfun$documentHighlights$1(MetalsLanguageServer.scala:1335)
        at scala.meta.internal.metals.CancelTokens$.$anonfun$apply$2(CancelTokens.scala:26)
        at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
        at scala.util.Success.$anonfun$map$1(Try.scala:255)
        at scala.util.Success.map(Try.scala:213)
        at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
        at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
        at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
        ... 4 more
```

This happens with both Bloop and sbt BSP.


**To Reproduce**

Steps to reproduce the behavior:

1. Clone https://github.com/eed3si9n/expecty
2. Open `expecty/src/main/scala-3/com/eed3si9n/expecty/RecorderMacro.scala` and import project
3. See errors

**Expected behavior**

No errors in sight

**Installation:**

- Operating system: macOS
- Editor: NeoVim
- Metals version: v 0.10.4+87-7b502251-SNAPSHOT
--

**Search terms**

Scala 3 macros

--- a/metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala
+++ b/metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala
@@ -11,6 +11,7 @@ import scala.meta.internal.metals.Buffers
 import scala.meta.internal.metals.BuildTargets
 import scala.meta.internal.metals.DefinitionProvider
 import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.ScalaVersionSelector
 import scala.meta.internal.mtags.GlobalSymbolIndex
 import scala.meta.internal.mtags.Mtags
 import scala.meta.internal.mtags.Semanticdbs
@@ -40,7 +41,8 @@ final class ImplementationProvider(
     buildTargets: BuildTargets,
     buffer: Buffers,
     definitionProvider: DefinitionProvider,
-    trees: Trees
+    trees: Trees,
+    scalaVersionSelector: ScalaVersionSelector
 ) {
   import ImplementationProvider._
 
@@ -432,6 +434,23 @@ final class ImplementationProvider(
     classInfo.map(inf => dealiasClass(inf, findSymbol))
   }
 
+  private def findDefOccurrence(
+      semanticDb: TextDocument,
+      symbol: String,
+      source: AbsolutePath
+  ): Option[SymbolOccurrence] = {
+    def isDefinitionOccurrence(occ: SymbolOccurrence) =
+      occ.role.isDefinition && occ.symbol == symbol
+
+    semanticDb.occurrences
+      .find(isDefinitionOccurrence)
+      .orElse(
+        Mtags
+          .allToplevels(source.toInput, scalaVersionSelector.getDialect(source))
+          .occurrences
+          .find(isDefinitionOccurrence)
+      )
+  }
 }
 
 object ImplementationProvider {
@@ -534,24 +553,6 @@ object ImplementationProvider {
       case _ =>
         Seq.empty
     }
-  }
-
-  def findDefOccurrence(
-      semanticDb: TextDocument,
-      symbol: String,
-      source: AbsolutePath
-  ): Option[SymbolOccurrence] = {
-    def isDefinitionOccurrence(occ: SymbolOccurrence) =
-      occ.role.isDefinition && occ.symbol == symbol
-
-    semanticDb.occurrences
-      .find(isDefinitionOccurrence)
-      .orElse(
-        Mtags
-          .allToplevels(source.toInput)
-          .occurrences
-          .find(isDefinitionOccurrence)
-      )
   }
 
   def isClassLike(info: SymbolInformation): Boolean =

--- a/metals/src/main/scala/scala/meta/internal/metals/BuildTargets.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/BuildTargets.scala
@@ -10,13 +10,11 @@ import scala.annotation.tailrec
 import scala.collection.concurrent.TrieMap
 import scala.collection.mutable
 import scala.collection.mutable.ListBuffer
-import scala.util.Try
 import scala.util.control.NonFatal
 
 import scala.meta.internal.io.PathIO
 import scala.meta.internal.metals.MetalsEnrichments._
 import scala.meta.internal.metals.ammonite.Ammonite
-import scala.meta.internal.mtags.Mtags
 import scala.meta.internal.mtags.Symbol
 import scala.meta.io.AbsolutePath
 
@@ -290,13 +288,8 @@ final class BuildTargets(
    * When encountering an unknown `readonly/` file we do the following steps to
    * infer what build target it belongs to:
    *
-   * - if file is in `.metals/readonly/dependencies/${source-jar-name}`
-   *    - find the build targets that have a sourceDependency with that name
-   * - if it isn't
-   *    - extract toplevel symbol definitions from the source code.
-   *    - find a jar file from any classfile that defines one of the toplevel
-   *      symbols.
-   *    - find the build target which has that jar file in it's classpath.
+   * - check if file is in `.metals/readonly/dependencies/${source-jar-name}`
+   * - find the build targets that have a sourceDependency with that name
    *
    * Otherwise if it's a jar file we find a build target it belongs to.
    *
@@ -318,7 +311,7 @@ final class BuildTargets(
                   ids.head
               }
           case _ =>
-            Try(unsafeInferBuildTarget(source)).getOrElse(None)
+            None
         }
       case None =>
         // else it can be a source file inside a jar
@@ -370,21 +363,6 @@ final class BuildTargets(
       .map(_.getId())
   }
 
-  private def unsafeInferBuildTarget(
-      source: AbsolutePath
-  ): Option[BuildTargetIdentifier] = {
-    val input = source.toInput
-    val toplevels = Mtags
-      .allToplevels(input)
-      .occurrences
-      .map(occ => Symbol(occ.symbol).toplevel)
-      .toSet
-    inferBuildTarget(toplevels).map { inferred =>
-      // Persist inferred result to avoid re-computing it again and again.
-      tables.foreach(_.dependencySources.setBuildTarget(source, inferred.id))
-      inferred.id
-    }
-  }
   case class InferredBuildTarget(
       jar: AbsolutePath,
       symbol: String,

--- a/metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala
@@ -52,7 +52,8 @@ final class DefinitionProvider(
     compilers: () => Compilers,
     remote: RemoteLanguageServer,
     trees: Trees,
-    buildTargets: BuildTargets
+    buildTargets: BuildTargets,
+    scalaVersionSelector: ScalaVersionSelector
 )(implicit ec: ExecutionContext) {
 
   val destinationProvider = new DestinationProvider(
@@ -210,7 +211,7 @@ final class DefinitionProvider(
 
   private def fromMtags(source: AbsolutePath, dirtyPos: Position) = {
     Mtags
-      .allToplevels(source.toInput)
+      .allToplevels(source.toInput, scalaVersionSelector.getDialect(source))
       .occurrences
       .find(_.encloses(dirtyPos))
   }

--- a/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
@@ -469,7 +469,8 @@ class MetalsLanguageServer(
           () => compilers,
           remote,
           trees,
-          buildTargets
+          buildTargets,
+          scalaVersionSelector
         )
         formattingProvider = new FormattingProvider(
           workspace,
@@ -503,7 +504,8 @@ class MetalsLanguageServer(
           buildTargets,
           buffers,
           definitionProvider,
-          trees
+          trees,
+          scalaVersionSelector
         )
 
         supermethods = new Supermethods(
@@ -1692,7 +1694,8 @@ class MetalsLanguageServer(
         val session = for {
           params <- debugSessionParams
           server <- debugProvider.start(
-            params
+            params,
+            scalaVersionSelector
           )
         } yield {
           cancelables.add(server)

--- a/metals/src/main/scala/scala/meta/internal/metals/ScalaVersionSelector.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/ScalaVersionSelector.scala
@@ -1,7 +1,9 @@
 package scala.meta.internal.metals
 
-import scala.meta.Dialect
+import scala.meta._
+import scala.meta.internal.metals.MetalsEnrichments._
 import scala.meta.internal.semver.SemVer
+import scala.meta.io.AbsolutePath
 
 class ScalaVersionSelector(
     userConfig: () => UserConfiguration,
@@ -48,5 +50,29 @@ class ScalaVersionSelector(
       fallbackScalaVersion(isAmmonite),
       includeSource3 = true
     )
+  }
+
+  def getDialect(path: AbsolutePath): Dialect = {
+
+    def dialectFromBuildTarget = buildTargets
+      .inverseSources(path)
+      .flatMap(id => buildTargets.scalaTarget(id))
+      .map(_.dialect)
+
+    Option(path.extension) match {
+      case Some("scala") =>
+        dialectFromBuildTarget.getOrElse(
+          fallbackDialect(isAmmonite = false)
+        )
+      case Some("sbt") => dialects.Sbt
+      case Some("sc") =>
+        // worksheets support Scala 3, but ammonite scripts do not
+        val dialect = dialectFromBuildTarget.getOrElse(
+          fallbackDialect(isAmmonite = path.isAmmoniteScript)
+        )
+        dialect
+          .copy(allowToplevelTerms = true, toplevelSeparator = "")
+      case _ => fallbackDialect(isAmmonite = false)
+    }
   }
 }

--- a/metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala
@@ -35,6 +35,7 @@ import scala.meta.internal.metals.MetalsBuildClient
 import scala.meta.internal.metals.MetalsEnrichments._
 import scala.meta.internal.metals.MetalsLanguageClient
 import scala.meta.internal.metals.MetalsStatusParams
+import scala.meta.internal.metals.ScalaVersionSelector
 import scala.meta.internal.metals.StacktraceAnalyzer
 import scala.meta.internal.metals.StatusBar
 import scala.meta.internal.metals.config.RunType
@@ -79,7 +80,8 @@ class DebugProvider(
   )
 
   def start(
-      parameters: b.DebugSessionParams
+      parameters: b.DebugSessionParams,
+      scalaVersionSelector: ScalaVersionSelector
   )(implicit ec: ExecutionContext): Future[DebugServer] = {
     Future.fromTry(parseSessionName(parameters)).flatMap { sessionName =>
       val inetAddress = InetAddress.getByName("127.0.0.1")
@@ -132,6 +134,7 @@ class DebugProvider(
             connectToServer,
             classFinder,
             stacktraceAnalyzer,
+            scalaVersionSelector,
             clientConfig.disableColorOutput()
           )
       }

--- a/metals/src/main/scala/scala/meta/internal/metals/debug/DebugProxy.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/debug/DebugProxy.scala
@@ -9,6 +9,7 @@ import scala.concurrent.Promise
 
 import scala.meta.internal.metals.Cancelable
 import scala.meta.internal.metals.GlobalTrace
+import scala.meta.internal.metals.ScalaVersionSelector
 import scala.meta.internal.metals.StacktraceAnalyzer
 import scala.meta.internal.metals.debug.DebugProtocol.ErrorOutputNotification
 import scala.meta.internal.metals.debug.DebugProtocol.InitializeRequest
@@ -30,6 +31,7 @@ private[debug] final class DebugProxy(
     server: ServerAdapter,
     classFinder: ClassFinder,
     stackTraceAnalyzer: StacktraceAnalyzer,
+    scalaVersionSelector: ScalaVersionSelector,
     stripColor: Boolean
 )(implicit ec: ExecutionContext) {
   private val exitStatus = Promise[ExitStatus]()
@@ -40,7 +42,12 @@ private[debug] final class DebugProxy(
   private val adapters = new MetalsDebugAdapters
 
   private val handleSetBreakpointsRequest =
-    new SetBreakpointsRequestHandler(server, adapters, classFinder)
+    new SetBreakpointsRequestHandler(
+      server,
+      adapters,
+      classFinder,
+      scalaVersionSelector
+    )
 
   lazy val listen: Future[ExitStatus] = {
     scribe.info(s"Starting debug proxy for [$sessionName]")
@@ -162,6 +169,7 @@ private[debug] object DebugProxy {
       connectToServer: () => Future[Socket],
       classFinder: ClassFinder,
       stacktraceAnalyzer: StacktraceAnalyzer,
+      scalaVersionSelector: ScalaVersionSelector,
       stripColor: Boolean
   )(implicit ec: ExecutionContext): Future[DebugProxy] = {
     for {
@@ -181,6 +189,7 @@ private[debug] object DebugProxy {
       server,
       classFinder,
       stacktraceAnalyzer,
+      scalaVersionSelector,
       stripColor
     )
   }

--- a/metals/src/main/scala/scala/meta/internal/metals/debug/SetBreakpointsRequestHandler.scala
+++ b/metals/src/main/scala/scala/meta/internal/metals/debug/SetBreakpointsRequestHandler.scala
@@ -5,6 +5,7 @@ import scala.concurrent.Future
 
 import scala.meta.internal.metals.JvmSignatures
 import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.ScalaVersionSelector
 import scala.meta.internal.mtags.Mtags
 import scala.meta.internal.parsing.ClassFinder
 import scala.meta.internal.semanticdb.Language
@@ -20,7 +21,8 @@ import org.eclipse.lsp4j.debug.SourceBreakpoint
 private[debug] final class SetBreakpointsRequestHandler(
     server: ServerAdapter,
     adapters: MetalsDebugAdapters,
-    classFinder: ClassFinder
+    classFinder: ClassFinder,
+    scalaVersionSelector: ScalaVersionSelector
 )(implicit ec: ExecutionContext) {
 
   private val previousBreakpointClassNames =
@@ -40,7 +42,10 @@ private[debug] final class SetBreakpointsRequestHandler(
     val symbols: Array[(SourceBreakpoint, Option[String])] =
       path.toLanguage match {
         case Language.JAVA =>
-          val topLevels = Mtags.allToplevels(path.toInput)
+          val topLevels = Mtags.allToplevels(
+            path.toInput,
+            scalaVersionSelector.getDialect(path)
+          )
           request.getBreakpoints.map { breakpoint =>
             val symbol = topLevels.occurrences.minBy(distanceFrom(breakpoint))
             (breakpoint, Option(JvmSignatures.toTypeSignature(symbol).value))

--- a/metals/src/main/scala/scala/meta/internal/parsing/Trees.scala
+++ b/metals/src/main/scala/scala/meta/internal/parsing/Trees.scala
@@ -32,7 +32,7 @@ final class Trees(
   def get(path: AbsolutePath): Option[Tree] =
     trees.get(path).orElse {
       // Fallback to parse without caching result.
-      parse(path, getDialect(path)).flatMap(_.toOption)
+      parse(path, scalaVersionSelector.getDialect(path)).flatMap(_.toOption)
     }
 
   def didClose(fileUri: AbsolutePath): Unit = {
@@ -80,7 +80,7 @@ final class Trees(
    * @return list of errors if the file failed to parse
    */
   def didChange(path: AbsolutePath): List[Diagnostic] = {
-    val dialect = getDialect(path)
+    val dialect = scalaVersionSelector.getDialect(path)
     parse(path, dialect) match {
       case Some(parsed) =>
         parsed match {
@@ -102,7 +102,7 @@ final class Trees(
   }
 
   def tokenized(input: inputs.Input.VirtualFile): Tokenized =
-    getDialect(AbsolutePath(input.path))(input).tokenize
+    scalaVersionSelector.getDialect(AbsolutePath(input.path))(input).tokenize
 
   private def parse(
       path: AbsolutePath,
@@ -113,32 +113,6 @@ final class Trees(
     } yield {
       val input = Input.VirtualFile(path.toString(), text)
       dialect(input).parse[Source]
-    }
-  }
-
-  private def getDialect(path: AbsolutePath): Dialect = {
-
-    def dialectFromBuildTarget = buildTargets
-      .inverseSources(path)
-      .flatMap(id => buildTargets.scalaTarget(id))
-      .map(_.dialect)
-
-    Option(path.extension) match {
-      case Some("scala") =>
-        dialectFromBuildTarget.getOrElse(
-          scalaVersionSelector.fallbackDialect(isAmmonite = false)
-        )
-      case Some("sbt") => dialects.Sbt
-      case Some("sc") =>
-        // worksheets support Scala 3, but ammonite scripts do not
-        val dialect = dialectFromBuildTarget.getOrElse(
-          scalaVersionSelector.fallbackDialect(isAmmonite =
-            path.isAmmoniteScript
-          )
-        )
-        dialect
-          .copy(allowToplevelTerms = true, toplevelSeparator = "")
-      case _ => scalaVersionSelector.fallbackDialect(isAmmonite = false)
     }
   }
 

--- a/metals/src/main/scala/scala/meta/internal/tvp/MetalsTreeViewProvider.scala
+++ b/metals/src/main/scala/scala/meta/internal/tvp/MetalsTreeViewProvider.scala
@@ -7,6 +7,7 @@ import java.util.concurrent.atomic.AtomicBoolean
 
 import scala.collection.concurrent.TrieMap
 
+import scala.meta.dialects
 import scala.meta.internal.metals.MetalsEnrichments._
 import scala.meta.internal.metals._
 import scala.meta.internal.mtags.GlobalSymbolIndex
@@ -267,7 +268,14 @@ class MetalsTreeViewProvider(
   ): Option[TreeViewNodeRevealResult] = {
     val input = path.toInput
     val occurrences =
-      Mtags.allToplevels(input).occurrences.filterNot(_.symbol.isPackage)
+      Mtags
+        .allToplevels(
+          input,
+          // TreeViewProvider doesn't work with Scala 3 - see #2859
+          dialects.Scala213
+        )
+        .occurrences
+        .filterNot(_.symbol.isPackage)
     if (occurrences.isEmpty) None
     else {
       val closestSymbol = occurrences.minBy { occ =>

--- a/mtags/src/main/scala/scala/meta/internal/mtags/Mtags.scala
+++ b/mtags/src/main/scala/scala/meta/internal/mtags/Mtags.scala
@@ -86,7 +86,7 @@ object Mtags {
 
   def allToplevels(
       input: Input.VirtualFile,
-      dialect: Dialect = dialects.Scala213
+      dialect: Dialect
   ): TextDocument = {
     input.toLanguage match {
       case Language.JAVA =>

--- a/tests/unit/src/test/scala/tests/WorkspaceSymbolExpectSuite.scala
+++ b/tests/unit/src/test/scala/tests/WorkspaceSymbolExpectSuite.scala
@@ -1,5 +1,6 @@
 package tests
 
+import scala.meta.dialects
 import scala.meta.internal.inputs._
 import scala.meta.internal.metals.WorkspaceSymbolProvider
 import scala.meta.internal.mtags.Mtags
@@ -14,7 +15,7 @@ class WorkspaceSymbolExpectSuite
         file,
         { () =>
           val input = file.input
-          val mtags0 = Mtags.allToplevels(input)
+          val mtags0 = Mtags.allToplevels(input, dialects.Scala213)
           val symtab0 = mtags0.symbols.map(i => i.symbol -> i).toMap
           val mtags = mtags0.copy(
             occurrences = mtags0.occurrences.filter { occ =>

